rust/src/dataset/scanner.rs

```diff
@@ -378,26 +378,85 @@ impl Scanner {
             let knn_node = self.ann(q, &index)?; // score, _rowid
             let with_vector = self.dataset.schema().project(&[&q.column])?;
             let knn_node_with_vector = self.take(knn_node, &with_vector)?;
-            let knn_node = if q.refine_factor.is_some() {
-                self.flat_knn(knn_node_with_vector, q)?
+            let mut knn_node = if q.refine_factor.is_some() {
+                self.flat_knn(vec![knn_node_with_vector], q)?
             } else {
                 knn_node_with_vector
             }; // vector, score, _rowid
+
+            // Check if we've created new versions since the index
+            let version = index.dataset_version;
+            if version != self.dataset.version().version {
+                // If we've added more rows, then we'll have new fragments
```

Comment:
could you extract this piece to a separate function?

Reply:
done